## Data Understanding

### Dataset Column Descriptions

| Column | Description |
|---|---|
| `mag` | Magnitude of the earthquake |
| `place` | Approximate location description (distance/direction from a known place) |
| `time` | Time the earthquake occurred (UTC) |
| `updated` | Time this record was last updated in USGS's database |
| `tz` | Timezone offset |
| `url` | Link to the USGS event page |
| `detail` | Link to the detailed GeoJSON feed for this specific event |
| `felt` | Number of "Did You Feel It?" reports submitted by the public — **Target Column** |
| `cdi` | Community Determined Intensity — max intensity reported by the public via DYFI |
| `mmi` | Modified Mercalli Intensity — estimated instrumental shaking intensity |
| `alert` | PAGER alert level (green/yellow/orange/red) estimating impact severity |
| `status` | Review status of the event (`automatic` or `reviewed`) |
| `tsunami` | Flag (0/1) indicating tsunami association (mainly oceanic events) |
| `sig` | A significance score combining magnitude, alerts, and felt reports |
| `net` | ID of the network that reported the event (e.g. `us`, `ak`, `pr`) |
| `code` | Identifying code assigned by the contributing network |
| `ids` | List of event IDs associated with this event across networks |
| `sources` | List of contributing networks |
| `types` | List of data product types available for this event |
| `nst` | Number of seismic stations used to locate the earthquake |
| `dmin` | Distance (degrees) from the epicenter to the nearest station |
| `rms` | Root-mean-square residual — indicates location quality/accuracy |
| `gap` | Largest azimuthal gap between stations (degrees) — smaller = more reliable location |
| `magType` | Method used to calculate magnitude (e.g. `ml`, `mb`, `mww`) |
| `type` | Type of seismic event (`earthquake`, `quarry blast`, etc.) |
| `title` | Auto-generated headline combining magnitude and place |
| `longitude` | Epicenter longitude |
| `latitude` | Epicenter latitude |
| `depth` | Depth of the earthquake below the surface (km) |

In [316]:
# Import the required libraries

import numpy as np , pandas as pd , plotly.express as px
import json


In [317]:
# Convert GeoJSON to DataFrame

def load_geojson(path):
    with open(path) as f:
        data = json.load(f)
    records = []
    for feature in data["features"]:
        row = dict(feature["properties"])
        row["longitude"], row["latitude"], row["depth"] = feature["geometry"]["coordinates"]
        records.append(row)
    return pd.DataFrame(records)

df1 = load_geojson("Dataset\\query_17_4_2025 to 14_12_2025.json")
df2 = load_geojson("Dataset\\query_16-12_2025 to 18_8_2026.json")

df = pd.concat([df1, df2], ignore_index=True)

df.to_csv("Dataset\\earthquakes_merged.csv", index=False)
print(df.shape)

(39113, 29)


In [318]:
pd.set_option('display.max_columns', 29)

In [319]:
# Data loading

df = pd.read_csv("Dataset\\earthquakes_merged.csv")
df

,mag,place,time,updated,tz,url,detail,felt,cdi,mmi,alert,status,tsunami,sig,net,code,ids,sources,types,nst,dmin,rms,gap,magType,type,title,longitude,latitude,depth
0,2.70,"109 km WNW of Yakutat, Alaska",1765756492747,1773181122040,NaN,https://earthquake.usgs.gov/earthquakes/eventp...,https://earthquake.usgs.gov/fdsnws/event/1/que...,NaN,NaN,NaN,NaN,reviewed,0,112,ak,2025younvr,",us6000ruvk,ak2025younvr,",",us,ak,",",origin,phase-data,",36.0,0.30000,0.90,146.0,ml,earthquake,"M 2.7 - 109 km WNW of Yakutat, Alaska",-141.4430,60.0150,13.100
1,4.30,"27 km W of Nemuro, Japan",1765755849253,1773181129040,NaN,https://earthquake.usgs.gov/earthquakes/eventp...,https://earthquake.usgs.gov/fdsnws/event/1/que...,NaN,NaN,NaN,NaN,reviewed,0,284,us,7000rn9l,",us7000rn9l,",",us,",",origin,phase-data,",34.0,2.00700,0.29,188.0,mb,earthquake,"M 4.3 - 27 km W of Nemuro, Japan",145.2408,43.3059,51.870
2,3.08,"6 km SW of Kenwood, CA",1765755518870,1773181122040,NaN,https://earthquake.usgs.gov/earthquakes/eventp...,https://earthquake.usgs.gov/fdsnws/event/1/que...,70.0,4.1,NaN,NaN,reviewed,0,175,nc,75279976,",nc75279976,us6000ruvg,",",nc,us,",",dyfi,focal-mechanism,moment-tensor,nearby-cit...",146.0,0.03365,0.12,16.0,mw,earthquake,"M 3.1 - 6 km SW of Kenwood, CA",-122.6045,38.3775,4.640
3,4.03,"6 km W of Glen Ellen, CA",1765755043210,1786923650579,NaN,https://earthquake.usgs.gov/earthquakes/eventp...,https://earthquake.usgs.gov/fdsnws/event/1/que...,1174.0,4.9,5.238,green,reviewed,1,740,nc,75279971,",ew1765755040,at00t7a9b8,us6000ruvf,nc75279971,",",ew,at,us,nc,",",dyfi,event-sequence,focal-mechanism,impact-li...",181.0,NaN,0.16,16.0,mw,earthquake,"M 4.0 - 6 km W of Glen Ellen, CA",-122.6040,38.3730,3.130
4,4.00,"169 km S of Wonosari, Indonesia",1765754774697,1773181129040,NaN,https://earthquake.usgs.gov/earthquakes/eventp...,https://earthquake.usgs.gov/fdsnws/event/1/que...,NaN,NaN,NaN,NaN,reviewed,0,246,us,7000rn9n,",us7000rn9n,",",us,",",origin,phase-data,",18.0,2.42300,0.96,93.0,mb,earthquake,"M 4.0 - 169 km S of Wonosari, Indonesia",110.4116,-9.4878,10.000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
39108,2.90,"58 km ENE of Susitna North, Alaska",1765846550322,1773181123040,NaN,https://earthquake.usgs.gov/earthquakes/eventp...,https://earthquake.usgs.gov/fdsnws/event/1/que...,NaN,NaN,NaN,NaN,reviewed,0,129,ak,2025yqshgt,",us6000rv5w,ak2025yqshgt,",",us,ak,",",origin,phase-data,",96.0,0.50000,0.90,22.0,ml,earthquake,"M 2.9 - 58 km ENE of Susitna North, Alaska",-148.8040,62.3460,26.900
39109,4.20,"71 km N of Port-Vila, Vanuatu",1765845780974,1773181129040,NaN,https://earthquake.usgs.gov/earthquakes/eventp...,https://earthquake.usgs.gov/fdsnws/event/1/que...,NaN,NaN,NaN,NaN,reviewed,0,271,us,7000rnnb,",us7000rnnb,",",us,",",origin,phase-data,",19.0,2.01200,0.82,83.0,mb,earthquake,"M 4.2 - 71 km N of Port-Vila, Vanuatu",168.4113,-17.1006,215.635
39110,2.50,"100 km N of Yakutat, Alaska",1765845735754,1768524138077,NaN,https://earthquake.usgs.gov/earthquakes/eventp...,https://earthquake.usgs.gov/fdsnws/event/1/que...,NaN,NaN,NaN,NaN,reviewed,0,96,ak,2025yqrvpi,",ak2025yqrvpi,",",ak,",",origin,phase-data,",18.0,0.40000,0.90,81.0,ml,earthquake,"M 2.5 - 100 km N of Yakutat, Alaska",-139.8570,60.4480,1.700
39111,3.10,"32 km SSE of Unalaska, Alaska",1765845465903,1773181130040,NaN,https://earthquake.usgs.gov/earthquakes/eventp...,https://earthquake.usgs.gov/fdsnws/event/1/que...,NaN,NaN,NaN,NaN,reviewed,0,148,us,7000rpsy,",ak2025yqrrrt,us7000rpsy,",",ak,us,",",origin,phase-data,",34.0,0.25900,0.34,199.0,ml,earthquake,"M 3.1 - 32 km SSE of Unalaska, Alaska",-166.3976,53.5945,65.389


## Data Exploration

Check for :

- Types
- NANs
- Duplicated Values
- Summary Statistics for (Numerical & Categorical)

In [320]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 39113 entries, 0 to 39112
Data columns (total 29 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   mag        39113 non-null  float64
 1   place      39113 non-null  str    
 2   time       39113 non-null  int64  
 3   updated    39113 non-null  int64  
 4   tz         0 non-null      float64
 5   url        39113 non-null  str    
 6   detail     39113 non-null  str    
 7   felt       5046 non-null   float64
 8   cdi        5046 non-null   float64
 9   mmi        3323 non-null   float64
 10  alert      1270 non-null   str    
 11  status     39113 non-null  str    
 12  tsunami    39113 non-null  int64  
 13  sig        39113 non-null  int64  
 14  net        39113 non-null  str    
 15  code       39113 non-null  str    
 16  ids        39113 non-null  str    
 17  sources    39113 non-null  str    
 18  types      39113 non-null  str    
 19  nst        38284 non-null  float64
 20  dmin       38270 

### Why we converted the `time` column?
`time` originally comes as a Unix timestamp (raw milliseconds), which isn't 
usable for analysis on its own. We converted it to a proper datetime so we 
can later engineer features from it (like month, day, or hour) and generate 
insights — for example, plotting how earthquake frequency changes month by 
month across 2025.


In [321]:
df['time'] = pd.to_datetime(df['time'], unit='ms')
df['time']

0       2025-12-14 23:54:52.747
1       2025-12-14 23:44:09.253
2       2025-12-14 23:38:38.870
3       2025-12-14 23:30:43.210
4       2025-12-14 23:26:14.697
                  ...          
39108   2025-12-16 00:55:50.322
39109   2025-12-16 00:43:00.974
39110   2025-12-16 00:42:15.754
39111   2025-12-16 00:37:45.903
39112   2025-12-16 00:30:40.757
Name: time, Length: 39113, dtype: datetime64[ms]

In [322]:
df.isnull().mean() * 100

mag            0.000000
place          0.000000
time           0.000000
updated        0.000000
tz           100.000000
url            0.000000
detail         0.000000
felt          87.098919
cdi           87.098919
mmi           91.504103
alert         96.752998
status         0.000000
tsunami        0.000000
sig            0.000000
net            0.000000
code           0.000000
ids            0.000000
sources        0.000000
types          0.000000
nst            2.119500
dmin           2.155294
rms            0.002557
gap            2.134840
magType        0.000000
type           0.000000
title          0.000000
longitude      0.000000
latitude       0.000000
depth          0.000000
dtype: float64

### Handling missing [administrative columns]
`tz` is 100% empty and provides no information, so it was dropped along 
with other purely administrative columns (`url`, `detail`, `title`, `ids`, 
`sources`, `types`, `code`, `updated`) that don't represent physical 
earthquake properties (metadata about the record itself).
They can't help the model learn anything meaningful

In [323]:
dropped_columns = ['tz', 'url', 'detail', 'title', 'ids', 'sources', 'types', 'code', 'updated']

df.drop(columns=dropped_columns, inplace=True)

In [324]:
df.duplicated().sum()

np.int64(0)

`cdi`, `mmi`, and `alert` are kept for now to visualize their relationship 
with `felt` before excluding them from training due to data leakage. 

`alert` is heavily missing (96.7%) and will also be excluded — not for 
leakage, but because too little real data exists to use it reliably.

`nst`, `dmin`, `rms`, `gap` drop NANS as it's a small percentage.


In [325]:
drop_nans = ['nst' , 'dmin' , 'rms' , 'gap']

df.dropna(subset=drop_nans, inplace=True , ignore_index=True)

- Summary Statistics for Numerical Columns

In [326]:
df.describe()

,mag,time,felt,cdi,mmi,tsunami,sig,nst,dmin,rms,gap,longitude,latitude,depth
count,38270.000000,38270,4846.000000,4846.000000,3190.000000,38270.000000,38270.000000,38270.000000,38270.000000,38270.000000,38270.000000,38270.000000,38270.000000,38270.000000
mean,3.904697,2025-12-17 12:01:14.882000,79.106892,2.861597,3.219131,0.003188,248.321087,44.190672,2.395683,0.664672,134.421324,-6.996754,25.078517,56.119098
min,2.500000,2025-04-18 00:01:58.239000,0.000000,0.000000,0.000000,0.000000,96.000000,0.000000,0.000000,0.000000,9.000000,-179.999700,-84.493200,-3.410000
25%,3.000000,2025-08-06 08:19:14.256000,1.000000,2.000000,2.250000,0.000000,142.000000,20.000000,0.500000,0.480000,80.000000,-140.101675,2.298975,10.000000
50%,4.200000,2025-12-14 18:46:31.279000,2.000000,2.700000,3.327000,0.000000,271.000000,31.000000,1.318000,0.670000,127.000000,-64.872900,34.879450,17.151000
75%,4.500000,2026-04-16 02:43:25.334000,11.000000,3.700000,4.106750,0.000000,312.000000,55.000000,2.722000,0.850000,184.000000,140.711325,51.874675,53.000000
max,8.800000,2026-08-18 03:33:08.783000,45935.000000,9.100000,9.393000,1.000000,2790.000000,566.000000,59.035000,3.400000,359.000000,179.999400,87.081500,683.578000
std,0.858810,NaN,1013.416891,1.365795,1.521202,0.056372,110.103919,38.168152,3.868301,0.290318,68.009367,135.279563,30.952991,103.127470


- Summary Statistics for Categorical Columns

In [327]:
df.describe(include='string')

,place,alert,status,net,magType,type
count,38270,1247,38270,38270,38270,38270
unique,23376,4,2,15,9,6
top,south of the Fiji Islands,green,reviewed,us,mb,earthquake
freq,433,1191,38131,30006,21324,37989


## Data Cleaning & Visualization 

In [328]:
num_cols = df.select_dtypes(include= 'number').columns
num_cols

Index(['mag', 'felt', 'cdi', 'mmi', 'tsunami', 'sig', 'nst', 'dmin', 'rms',
       'gap', 'longitude', 'latitude', 'depth'],
      dtype='str')

In [329]:
for col in num_cols:
    fig1 = px.histogram(df, x=col, nbins=50, title=f'Histogram of {col}' , color_discrete_sequence=[ "#3B5E64"])
    fig2 = px.box(df, x=col, title=f'Box Plot of {col}' , color_discrete_sequence= ["#3B5E64"] )
    fig1.show()
    fig2.show()

In [330]:
cat_cols = df.select_dtypes(include= 'str').columns
cat_cols

Index(['place', 'alert', 'status', 'net', 'magType', 'type'], dtype='str')

In [331]:
for col in cat_cols:
    print(f'Column :' , col)
    print(f'Number of Unique Values :' ,df[col].nunique())
    print(f'Unique Values :' ,df[col].unique())
    print('-' * 50)

Column : place
Number of Unique Values : 23376
Unique Values : <ArrowStringArray>
[                    '109 km WNW of Yakutat, Alaska',
                          '27 km W of Nemuro, Japan',
                            '6 km SW of Kenwood, CA',
                   '169 km S of Wonosari, Indonesia',
 '50 km WNW of San Antonio de los Cobres, Argentina',
                        '35 km WNW of Ester, Alaska',
                        '29 km E of Wright, Wyoming',
                            '62 km E of Noda, Japan',
                     '10 km N of Garden City, Texas',
                        '91 km N of Yakutat, Alaska',
 ...
                '210 km WSW of Pistol River, Oregon',
                   '135 km SSE of King Cove, Alaska',
                '15 km SSW of El Rosario, Nicaragua',
              '2 km SW of G. L. Garcia, Puerto Rico',
                         '29 km S of Ōfunato, Japan',
                    '30 km W of Rincón, Puerto Rico',
                         '61 km NE of Calama, Chi

In [332]:
df.drop('alert' , axis=1 , inplace=True)

- Checking The Correlation between `felt` and [`cdi` , `mmi`]
- corr() --> take not null values only.

In [333]:
corr = df[['felt', 'mmi', 'cdi']].corr()

fig = px.imshow(
    corr,
    text_auto=True,
    color_continuous_scale='RdBu_r',
    title='Correlation Heatmap: Felt, MMI, and CDI'
)

fig.show()

### Why `cdi` and `mmi` are excluded from the model?
A crosstab of `felt.notna()` vs `cdi.notna()` showed a perfect 100% match: 
every row with `felt` also has `cdi`, and vice versa, with zero exceptions 
— meaning `cdi`'s mere presence directly reveals the target. This is 
confirmed data leakage.

In [334]:
print(pd.crosstab(df['felt'].notna(), df['cdi'].notna()))

cdi    False  True 
felt               
False  33424      0
True       0   4846


`mmi` showed a weaker pattern: it doesn't align perfectly with `felt` 
(some quakes have `mmi` without `felt`, and vice versa), so it isn't a 
guaranteed leak. but also it's better to exclude it from data. 

In [335]:
print(pd.crosstab(df['felt'].notna(), df['mmi'].notna()))

mmi    False  True 
felt               
False  31463   1961
True    3617   1229


In [336]:
excluded_cols =['mmi', 'cdi']
df.drop(columns= excluded_cols , inplace= True)

- Handling Target Column (`felt`)

`felt` originally holds the number of "Did You Feel It?" reports, with 
missing values (`NaN`) where no one reported feeling the earthquake. We 
converted it into a binary target, `reported`:
- `NaN` or `0` → `0` (not reported / not felt)
- Any value greater than `0` → `1` (reported / felt)

In [337]:
df['reported'] = (df['felt'] > 0).astype(int)

In [338]:
df['reported'].isna().sum()

np.int64(0)

In [339]:
df.drop('felt' , axis=1 , inplace= True)

In [340]:
df.head()

,mag,place,time,status,tsunami,sig,net,nst,dmin,rms,gap,magType,type,longitude,latitude,depth,reported
0,2.70,"109 km WNW of Yakutat, Alaska",2025-12-14 23:54:52.747,reviewed,0,112,ak,36.0,0.30000,0.90,146.0,ml,earthquake,-141.4430,60.0150,13.100,0
1,4.30,"27 km W of Nemuro, Japan",2025-12-14 23:44:09.253,reviewed,0,284,us,34.0,2.00700,0.29,188.0,mb,earthquake,145.2408,43.3059,51.870,0
2,3.08,"6 km SW of Kenwood, CA",2025-12-14 23:38:38.870,reviewed,0,175,nc,146.0,0.03365,0.12,16.0,mw,earthquake,-122.6045,38.3775,4.640,1
3,4.00,"169 km S of Wonosari, Indonesia",2025-12-14 23:26:14.697,reviewed,0,246,us,18.0,2.42300,0.96,93.0,mb,earthquake,110.4116,-9.4878,10.000,0
4,4.30,"50 km WNW of San Antonio de los Cobres, Argentina",2025-12-14 23:24:04.484,reviewed,0,284,us,40.0,1.71400,0.76,62.0,mb,earthquake,-66.7998,-24.1137,198.342,0


## Feature Engineering

**Time-based features** — extracted from `time` since the raw datetime 
isn't directly usable by a model:
- `hour` — hour of day (0–23)
- `month` — month (1–12)
- `day_of_week` — day of week (0–6)

In [341]:
df['hour'] = df['time'].dt.hour
df['month'] = df['time'].dt.month
df['day_of_week'] = df['time'].dt.dayofweek
df.drop('time' , axis=1 , inplace=True)

`mag_depth_ratio` vs. `sig`

**Why it matters:** a magnitude-5 earthquake at 5 km depth is far more 
likely to be felt than a magnitude-5 earthquake at 500 km depth, even 
though they share the same magnitude. `mag` alone doesn't capture this 
difference, but `mag_depth_ratio` summarizes it in one number — a higher 
ratio means shallower and stronger, and therefore more likely to be felt.

`sig` is a ready-made score calculated by USGS using an internal formula 
that factors in magnitude, alerts, and actual public "felt" reports — so 
it's partly built from human reaction data.

In [342]:
df['mag_depth_ratio'] = df['mag'] / (df['depth'] + 1)

`place` is free text and not directly usable, so we extract one signal: 
does it contain "km" (near a populated place) or not (remote region).


In [343]:
df['is_remote'] = ~df['place'].str.contains('km', case=False, na=False)
df.drop('place', axis=1 , inplace=True) 

In [344]:
df.isna().sum()

mag                0
status             0
tsunami            0
sig                0
net                0
nst                0
dmin               0
rms                0
gap                0
magType            0
type               0
longitude          0
latitude           0
depth              0
reported           0
hour               0
month              0
day_of_week        0
mag_depth_ratio    0
is_remote          0
dtype: int64

In [345]:
df.head()

,mag,status,tsunami,sig,net,nst,dmin,rms,gap,magType,type,longitude,latitude,depth,reported,hour,month,day_of_week,mag_depth_ratio,is_remote
0,2.70,reviewed,0,112,ak,36.0,0.30000,0.90,146.0,ml,earthquake,-141.4430,60.0150,13.100,0,23,12,6,0.191489,False
1,4.30,reviewed,0,284,us,34.0,2.00700,0.29,188.0,mb,earthquake,145.2408,43.3059,51.870,0,23,12,6,0.081332,False
2,3.08,reviewed,0,175,nc,146.0,0.03365,0.12,16.0,mw,earthquake,-122.6045,38.3775,4.640,1,23,12,6,0.546099,False
3,4.00,reviewed,0,246,us,18.0,2.42300,0.96,93.0,mb,earthquake,110.4116,-9.4878,10.000,0,23,12,6,0.363636,False
4,4.30,reviewed,0,284,us,40.0,1.71400,0.76,62.0,mb,earthquake,-66.7998,-24.1137,198.342,0,23,12,6,0.021571,False


In [346]:
# Only for Visualization, will be excluded from model's training data.
df['reported_label'] = df['reported'].map({0: 'Not Felt', 1: 'Felt'})

## Data Analysis (For Target Column with other Input Features)

### What Makes an Earthquake Get Felt?

- Do stronger earthquakes (magnitude) get felt more often?

In [347]:
plot_df = df.groupby('reported_label')['mag'].mean().round(2).reset_index()
plot_df

,reported_label,mag
0,Felt,3.92
1,Not Felt,3.90


In [348]:
px.bar(plot_df, x='reported_label', y='mag', title='Magnitude Distribution: Felt vs Not Felt' ,
        color_discrete_sequence=[ "#3B5E64"], labels= {'reported_label' : 'Reported Label' , 'mag' : 'Magnitude'},
         text_auto= True )

### Insight 1 :

* Magnitude value itself doesn't reflect a useful indicator about the earthquakes, for that reason we get the mag_depth ratio.

##### ----------------------------------------------------------------------------------------------------------------------

- Does earthquake depth affect the likelihood of being felt?

In [349]:
plot_df = df.groupby('reported_label')['depth'].mean().round(2).reset_index()
plot_df

,reported_label,depth
0,Felt,31.41
1,Not Felt,59.54


In [350]:
px.bar(plot_df, x='reported_label', y='depth', title='Depth Distribution: Felt vs Not Felt' ,
        color_discrete_sequence=[ "#3B5E64"], labels= {'reported_label' : 'Reported Label' , 'depth' : 'Depth'},
         text_auto= True )

### Insight 2 :

- Deepest earthquakes mostly not felt by humans.

##### ----------------------------------------------------------------------------------------------------------------------


- Are earthquakes near populated areas felt more often than remote/oceanic ones?

In [351]:
plot_df = df.groupby('is_remote')['reported'].mean().reset_index(name= 'Average Reporting Rate')

plot_df['Average Reporting Rate']  = (plot_df['Average Reporting Rate'] * 100).round(2)

plot_df

,is_remote,Average Reporting Rate
0,False,13.57
1,True,1.57


In [352]:
px.bar(plot_df, x='is_remote', y='Average Reporting Rate', title='Location Type Distribution' ,
        color_discrete_sequence=[ "#3B5E64"], labels= {'is_remote' : 'Location Flag'},
         text_auto= True )

### Insight 3 :

- The nearest areas earthquakes exist, the more chance that humans felt, 13.6% for near locations vs. 1.6% for remote ones (~9x difference)
##### ----------------------------------------------------------------------------------------------------------------------

- Does the time of day or month affect reporting?

In [353]:
# Convert 'hour' into meaningful values due to the analysis view.

def hour_to_label(h):
    if h == 0:
        return "12 AM"
    elif h < 12:
        return f"{h} AM"
    elif h == 12:
        return "12 PM"
    else:
        return f"{h - 12} PM"

df['hour_label'] = df['hour'].apply(hour_to_label)

In [354]:
hour_order = [hour_to_label(h) for h in range(24)] # sorting 

plot_df = df.groupby('hour_label')['reported'].sum().reset_index()
plot_df

,hour_label,reported
0,1 AM,186
1,1 PM,169
2,10 AM,195
3,10 PM,187
4,11 AM,188
5,11 PM,185
6,12 AM,194
7,12 PM,179
8,2 AM,213
9,2 PM,184


In [355]:
px.bar(
    plot_df,
    x='hour_label',
    y='reported',
    title='Reporting Rate by Hour of Day',
    color_discrete_sequence=["#3B5E64"],
    labels={'hour_label': 'Hour of Day', 'reported': 'Reporting Rate'},
    text_auto=True,        
    category_orders={'hour_label': hour_order ,
                     }  
).update_xaxes(categoryorder='max descending')

In [356]:
month_map = {
    1: 'January', 2: 'February', 3: 'March', 4: 'April',
    5: 'May', 6: 'June', 7: 'July', 8: 'August',
    9: 'September', 10: 'October', 11: 'November', 12: 'December'
}

df['month_name'] = df['month'].map(month_map)

In [357]:
plot_df = df.groupby('month_name')['reported'].mean() .reset_index(name='Average Reporting Rate').sort_values(ascending=False , by='Average Reporting Rate')
plot_df['Average Reporting Rate'] = round(plot_df['Average Reporting Rate'] * 100 , 2)
plot_df

,month_name,Average Reporting Rate
8,May,15.43
10,October,14.01
4,January,13.86
0,April,13.37
6,June,12.88
11,September,12.84
9,November,12.80
3,February,12.78
7,March,11.86
1,August,11.16


In [358]:
# month_order = list(month_map.values())

px.line(
    plot_df,
    x='month_name',
    y='Average Reporting Rate',
    title='Average Reporting Rate by Month',
    markers=True,                          
    labels={'month_name': 'Month'},
    # category_orders={'month_name': month_order}
)

### Insight 4 :

- Reprorted Earthquakes mostly occured in the 'AM' period.
- 'May' is the most month receives earthquakes reports , the other side, 'December' is the lowest month
##### ----------------------------------------------------------------------------------------------------------------------

- Are tsunami-associated earthquakes felt differently than others?

In [359]:
plot_df = df.groupby('tsunami')['reported'].mean().reset_index(name= 'Average Reporting Rate')
plot_df['Average Reporting Rate']  = round(plot_df['Average Reporting Rate'] * 100 , 3)
plot_df

,tsunami,Average Reporting Rate
0,0,11.995
1,1,66.393


In [360]:
px.bar(
    plot_df,
    x='tsunami',
    y='Average Reporting Rate',
    title='Reporting Rate by Tsunami',
    color_discrete_sequence=["#3B5E64"],
    labels={'tsunami': 'Tsunami'},
    text_auto=True,          
)

### Insight 5 :
- There is a strong relationship between the occurrence of earthquakes and tsunamis.
##### ----------------------------------------------------------------------------------------------------------------------

In [361]:
# exclude the analysis columns (labels) as it already exists in the numerical form.

excluded_cols = ['reported_label', 'hour_label', 'month_name']

df.drop(columns= excluded_cols , inplace=True)

In [362]:
df.columns

Index(['mag', 'status', 'tsunami', 'sig', 'net', 'nst', 'dmin', 'rms', 'gap',
       'magType', 'type', 'longitude', 'latitude', 'depth', 'reported', 'hour',
       'month', 'day_of_week', 'mag_depth_ratio', 'is_remote'],
      dtype='str')

In [363]:
# df.to_csv('earthquakes_cleaned.csv' , index= False)